tools/db-checksum/CheckpointerDBChecksum.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE BangPatterns       #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections      #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module CheckpointerDBChecksum where
+
+import Control.Monad.Reader
+import Control.Exception.Safe (tryAny)
+
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import Data.ByteString.Builder
+import Data.Int
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Serialize
+import Data.Text (pack, Text)
+
+import Database.SQLite3.Direct as SQ3
+
+import GHC.Generics
+import GHC.IO.Handle
+
+import System.IO
+
+-- pact imports
+import Pact.Types.SQLite
+
+-- chainweb imports
+import Chainweb.BlockHash
+import Chainweb.BlockHeader
+import Chainweb.Pact.Backend.Types
+import Chainweb.Pact.Backend.Utils hiding (callDb)
+import Chainweb.Pact.Service.Types
+import Chainweb.Version
+
+main :: IO ()
+main = putStrLn "Yay, compiles!"
+
+
+type TableName = ByteString
+type TableContents = ByteString
+
+-- this will produce both the raw bytes for all of the tables concatenated
+-- together, and the raw bytes of each individual table (this is over a single chain)
+work :: Args -> FilePath -> BlockHeight -> IO (Builder, Map TableName TableContents)
+work _args sqlitefile limit = withSQLiteConnection sqlitefile chainwebPragmas False (runReaderT go)
+  where
+    go = do
+        names <- getTableNames
+        callDb "getting rows" $ \db -> liftIO $ foldM (collect db) (mempty, mempty) names
+
+    collect db (entiredb,  m) name = do
+        rows <- getTblContents db name
+        let !bytes = encode rows
+        return ((mappend entiredb (byteString bytes)), (M.insert name bytes m))
+
+    getTblContents db = \case
+        "TransactionIndex" -> qry db tistmt [SInt $ fromIntegral limit] [RBlob, RInt]
+        "BlockHistory" -> qry db bhstmt [SInt $ fromIntegral limit] [RInt, RBlob, RInt]
+        "VersionedTableCreation" -> qry db vtcstmt [SInt $ fromIntegral limit] [RText, RInt]
+        "VersionedTableMutation" -> qry db vtmstmt [SInt $ fromIntegral limit] [RText, RInt]
+        t -> do
+          txid <- getEndingTxId limit db
+          qry db (usertablestmt t) [SInt txid] [RText, RInt, RBlob]
+
+    getEndingTxId :: BlockHeight -> Database -> IO Int64
+    getEndingTxId bh db = do
+        r <- qry db stmt [SInt $ fromIntegral bh] [RInt]
+        case r of
+          [[SInt txid]] -> return txid
+          _ -> error "Cannot find ending txid."
+      where
+        stmt = "SELECT endingtxid FROM BlockHistory WHERE blockheight = ?;"
```

Comment:
add `ORDER BY`

Reply:
Why would we need `ORDER BY` here? There can only be one result by design.